receiver/statsdreceiver/transport/packet_server.go

```diff
@@ -0,0 +1,131 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package transport // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/statsdreceiver/transport"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strings"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.uber.org/multierr"
+)
+
+type packetServer struct {
+	packetConn net.PacketConn
+	transport  Transport
+	address    string
+}
+
+// Ensure that Server is implemented on Packet transport based servers.
+var _ (Server) = (*packetServer)(nil)
+
+// NewPacketServer creates a transport.Server using transports based on packets.
+func NewPacketServer(transport Transport, address string) (Server, error) {
+	var psrv packetServer
+	var err error
+
+	if !transport.IsPacketTransport() {
+		return nil, ErrUnsupportedPacketTransport
+	}
+	if transport.IsUnixTransport() {
+		err := os.Remove(address)
+		if err != nil && !errors.Is(err, os.ErrNotExist) {
+			return nil, fmt.Errorf("removing socket failed: %w", err)
+		}
+	}
+
+	psrv.transport = transport
+	psrv.address = address
+	psrv.packetConn, err = net.ListenPacket(transport.String(), address)
+	if err != nil {
+		return nil, fmt.Errorf("starting to listen %s socket: %w", psrv.transport, err)
+	}
+
+	return &psrv, nil
+}
+
+// Start the server ready to receive metrics.
+func (psrv *packetServer) ListenAndServe(
+	nextConsumer consumer.Metrics,
+	reporter Reporter,
+	transferChan chan<- Metric,
+) error {
+	if nextConsumer == nil || reporter == nil {
+		return errNilListenAndServeParameters
+	}
+
+	buf := make([]byte, 65527) // max size for udp packet body (assuming ipv6)
+	for {
+		n, addr, err := psrv.packetConn.ReadFrom(buf)
+
+		if _, ok := psrv.packetConn.(*net.UnixConn); ok && addr == nil {
```

Comment:
why can't we rely on psrv.transport?

Reply:
Since it is used as metadata only, would make sense to move this check in handlePacket?

Reply:
Fair enough. Done.